### DISCUSSION TITLE
strconv: raise error if not a number

### DISCUSSION THREAD

**Discussion by TheBrainScrambler** [vlang/v#11684] [Ideas]
Hello, I've just discovered the language and have read the docs. Now I want to practice, so I'm writing a little guessing game where you enter a number between 1 and 100.
I need to convert the user's input from a string to an int. Reading the docs, I've found https://modules.vlang.io/strconv.html#atoi which does what I want. BUT ... not exactly.
If I enter some random characters in my program, `strconv.atoi` will just convert it to 0 instead of rising an error where it says that the string can't be converted.
I think this is incorrect behavior, the strconv methods should be Optionals, or at least provide methods which implements this mechanisms and are Optionals.

**TL;DR**
**Expected behavior:** If I do `strconv.atoi("blablah")` it should let us handle the error with an Optional
**Current behavior:** It just returns 0 ...

**Comment by Delta456** [comment]
What's your say on this? @penguindark

**Comment by medvednikov** [comment]
`number := os.input('enter number').int()`

**Comment by medvednikov** [comment]
But I agree, atoi() should return `?int`

**Comment by TheBrainScrambler** [comment]
"number := os.input('enter number').int()"
This also returns 0 if I enter characters instead of returning an error

**Comment by Delta456** [comment]
if we add an optional for so `str.int()` will return 0 or error if it's a not a valid number right? @medvednikov

**Comment by TheBrainScrambler** [comment]
@Delta456 No, it just returns 0 if the number is invalid, like `strconv.atoi`. Test yourself by running `number := os.input('enter number').int()` and printing `number`. Enter "Blablah" for example, it will print you 0.

**Comment by Delta456** [comment]
@TheBrainScrambler I meant if we will add an optional so what will it return.

**Comment by TheBrainScrambler** [comment]
@Delta456 I don't understand, the optional should return 0 if it isn't an error ? What ?

I also discovered by using the `cli` module that there is the same behavior: it will convert my entry to 0 if it isn't a valid number, instead of returning an error. It is probably using `strconv.atoi` or the `.int()` method

**Comment by OpenSourceAnarchist** [comment]
Yeah I'm for making pretty much all user-interaction functions return an optional type. There should be no overlap between valid returns (like 0 in this case) and invalid returns (also 0 in this case). This seems like the perfect use-case of optionals.

**Comment by penguindark** [comment]
@medvednikov @Delta456 
`atoi` in C act like: `This function returns the converted integral number as an int value. If no valid conversion could be performed, it returns zero`
In V it has the same behaviour.
Adding an optional can be  done but I'm not sure if it is needed.

**Comment by OpenSourceAnarchist** [comment]
@penguindark C does not have optionals, so returning 0 is the best solution within the constraints of the language. V has a feature that would lend itself perfectly for parsing "iffy" data (unknown strings) with the certainty that whatever is returned by the function, if not an error, is precisely what the developer wants. Without optionals, there is no way to know if the user typed 0 or if they typed nonsense.

**Comment by JalonSolov** [comment]
It would perhaps be better to create a different method with a new name, instead of changing the way atoi works.

The official documentation for the parameter to atoi (in C, C++, etc.) is:

C-string **beginning with** the representation of an integral number.

The key part (that I bolded) is "beginning with".

This also means that atoi will return 5 if you pass it a string of "5ive", instead of giving an error.

So having a separate _strict_ version of atoi would be ok, but changing the way atoi works would just make porting existing code harder.

**Comment by penguindark** [comment]
I agree with JalonSolov
Add another method can be the best solution.

**Comment by TheBrainScrambler** [comment]
I don't know if the `strconv.atoi` and/or `.int()` come from C or V. Maybe just keep atoi for C compatibility by mentionning that in the docs and just modify the int method ... It would also be interesting to know what the `cli` module uses to convert strings to int

**Comment by JalonSolov** [comment]
Looks like .int()... example line:

`return flag.value.int()`

**Comment by TheBrainScrambler** [comment]
Then I think the .int() method should be modified to return an Optional and keep strconv.atoi for C compatibility.

This should also be marked as a bug. I can't properly use the CLI library because if an user enter some string on an int flag, it won't raise an error and instead return 0. But 0 is also a valid entry so I can't check if the entry is 0 ...

**Comment by medvednikov** [comment]
Yeah good point

**Comment by mcastorina** [comment]
`strconv.atoi` returns `?int` (V 0.2.2 6e589e2)

Can this issue be closed?

**Reply by medvednikov** [reply]
Indeed it can now :)

**Comment by thomas-mangin** [comment]
As a newcomer to the language, I was extremely surprised to find that `.int()` or `.u32()` did not use optional and require an `or` block. I had to find this issue to figure out what to do to make sure that invalid input would not result in a zero value.

the introduction of `strconv.atoi` did allow for both use cases (thank you) but is IMHO un-intuitive. `strconv.atoi` should have followed the behaviour of C (return 0) when the built-in should have provided an optional and be "safe by default".

As this issue is not closed, perhaps the behaviour can still be changed and vfmt used to invert the use of one with the other. If so, atou32, and friends may need to be added to vlib to allow a full substitution.

**Comment by prantlf** [comment]
Indeed, `atoi` in C returns `0` in case of failure, which makes it useless for obtain numbers including `0`. Instead of `atoi`, `strtol` has been recommended for a long time. It still returns `0` in case of a failure, but it also sets `errno` to `EINVAL`.

So, `.int()` in V ended up like `atoi` in C - next to useless. The equivalent of `strtol` in C was named `atoi` in V. How ironic :-)

**Reply by Wajinn** [reply]
> Indeed, atoi in C returns 0 in case of failure, which makes it useless for obtain numbers including 0. Instead of atoi, strtol has been recommended for a long time. It still returns 0 in case of a failure, but it also sets errno to EINVAL.

I'm confused as to what the issue is.  The V way allows a lot of flexibility.  You can return whatever needed.

`println(strconv.atoi('blahblah') or {0}) // will give you 0`

```
fn main() {
	println(is_numeric('blahblah')) // return false
}

fn is_numeric(s string) bool {
	strconv.atoi(s) or {
		return false
	}
	return true
}
```

> So, .int() in V ended up like atoi in C - next to useless. 

I'm not understanding what's meant by that.  "int()" is still necessary or useful.  If the programmer is seeking to test the result or looking for a particular result, there are various ways that they can do so.

**Reply by prantlf** [reply]
> I'm confused as to what the issue is. 

The issue was fixed in 0.2.2 by making `strconv.atoi` return `?int`. It stayed open to help people finding the proper integer parsing method, I guess. That's how I got to it. You could convert it to a discussion and close it.

> I'm not understanding what's meant by that. "int()" is still necessary or useful.

You assume that parsing a string to an integer _without checking for parsing errors_ is useful. In what real-world scenario would you omit error handling?

Every language has some history. You may call this "flexibility", the others "confusion", because they don't see, why it should be in the language.

For example, Go offers `strconv.Atoi`, `strconv.ParseInt` and `fmt.Sscanf`, all of them support error handling. They learnt from the history of C and didn't introduce the flexibility/confusion.

**Reply by Wajinn** [reply]
> For example, Go offers `strconv.Atoi`, `strconv.ParseInt` and `fmt.Sscanf`, all of them support error handling. They learnt from the history of C and didn't introduce the flexibility/confusion.

V is also offering `strconv.atoi`, `strconv.parse_int`, and `os.input`.  

There wouldn't necessarily be any confusion, other than making assumptions or unfamiliarity from being new.  A programmer would/should/could consult [V's STL documentation](https://modules.vlang.io/).  Nothing is stopping the programmer from making a function to handle or address what is needed, or to simply not use `.int()`.  That's what I'm referring to, in terms of flexibility.

However, maybe having `!int` (and doing similar for the related) would be more acceptable for most.  If an intended behavior is to help with C compatibility, then debatably it should indicate so from its name or in the documentation.  At this point, those that were using `atoi` between V and C, would have already adjusted as it's `!int`.

**Reply by thomas-mangin** [reply]
`.int()` is the "built-in" way to convert a string to an int and can not be used safely.

Yes, the library offers other options - which are less intuitive than using `.int()` and require importing modules. Depending on the language you "come from", you will either know the C API or will try a type casting (IMO more intuitive).

The V documentation has toward its top
```
Strings can be easily converted to integers:

s := '42'
n := s.int() // 42
```

And much after, an example about errors using `strconv`

```
If a test function has an error return type, any propagated errors will fail the test:

import strconv

fn test_atoi() ? {
	assert strconv.atoi('1')? == 1
	assert strconv.atoi('one')? == 1 // test will fail
}
```

The example code for int() does not present the "zero behaviour". This should be corrected and the section also updated to provide some information about `strconv.atoi`, `strconv.parse_int`, and `os.input`.

Any experienced programmer will notice the silencing of errors when trying to catch it, but junior programmers may miss the code path and experience a runtime issue, which is what typed languages are supposed to help prevent.

As I have read multiple statements on Discord about the developer's wish to keep the language predictable and beginner friendly, there is probably still a need to do something about this API and this is why the issue is probably still open.

I will continue to think that the behaviour is the wrong way round. `atoi` should return zero like C and `int()` be converted to `int()?`.

Alternatively, V could have a concept of optional with a default value where if the error is not caught, the default value is provided. I suggested this a while back (and attempted a RFC) but it got no traction and much opposition.

**Reply by Wajinn** [reply]
Seems like those who were dealing with C compatibility, would have already encountered or adapted to `!` from V's `strconv.atoi` and `strconv.parse_int`.  And if they ran over to using `.int()`, then maybe split it into 2 new functions, where one gives zero and the other is optional.

**Reply by prantlf** [reply]
You could add a rule to `v vet`, which would warn about the usage of `.int()` and recommend replacing it with `strconv.atoi()`. It's avoid mistakes from the omitting the error handling.

**Reply by Wajinn** [reply]
At least updating the documentation (both the V docs and STL docs) to provide more clarity, seems like the best move, regardless of what else is decided later.  In the V docs, as well as links to `strconv` and `os`, possibly also mention `strconv.atoi`, `strconv.parse_int`, and `os.input` (or others) too.  This way, newbies know to look there.  In the STL docs, maybe add example usage for those mentioned.

**Reply by thomas-mangin** [reply]
@prantlf imho the built-in should be working so that users do not need any import and the library should have kept the C api. Just my view.

**Reply by medvednikov** [reply]
@Wajinn is correct. That was my goal: have a simple to use method that returns 0 on error, and a more advanced function that returns a Result.
